gcc/rust/resolve/rust-ast-resolve-type.cc

```diff
@@ -477,6 +477,27 @@ ResolveTypeToCanonicalPath::visit (AST::SliceType &type)
     }
 }
 
+void
+ResolveTypeToCanonicalPath::visit (AST::TraitObjectTypeOneBound &type)
+{
+  CanonicalPath path = CanonicalPath::create_empty ();
+  bool ok
+    = ResolveTypeToCanonicalPath::go (&type.get_trait_bound ().get_type_path (),
+				      path);
+  if (ok)
+    {
+      std::string slice_path = "<dyn " + path.get () + ">";
+      result = CanonicalPath::new_seg (type.get_node_id (), slice_path);
+    }
+}
+
+void
+ResolveTypeToCanonicalPath::visit (AST::TraitObjectType &type)
+{
+  // FIXME is this actually allowed? dyn A+B
+  gcc_unreachable ();
```

Comment:
Do you mean this?

```rust
trait Foo {}
trait Bar {}

fn takes_dyn_foo_bar(_: &(dyn Foo + Bar)) {} // invalid
fn takes_dyn_foo_send(_: &(dyn Foo + Send)) {} // okay
```

The error message for the first functions is as follows:

```rust
error[[E0225]](https://doc.rust-lang.org/stable/error-index.html#E0225): only auto traits can be used as additional traits in a trait object
 --> src/lib.rs:4:37
  |
4 | fn takes_dyn_foo_bar(_: &(dyn Foo + Bar)) {} // invalid
  |                               ---   ^^^ additional non-auto trait
  |                               |
  |                               first non-auto trait
  |
  = help: consider creating a new trait with all of these as supertraits and using that trait here instead: `trait NewTrait: Foo + Bar {}`
  = note: auto-traits like `Send` and `Sync` are traits that have special properties; for more information on them, visit <https://doc.rust-lang.org/reference/special-types-and-traits.html#auto-traits>
  ```
  
  So it's allowed in some cases :)

Reply:
ah ok thats interesting thats the same error for impl dyn A+B too.

Reply:
That error handling must be during type resolution not 100% sure why this is the case in Rust but there is definitely a reason. We have an open bug about handling super traits in dyn objects which i have been meaning to fix for ages i bet its got to do with that.